migration/migrator_newstate.go

```diff
@@ -1,200 +1,161 @@
 			}
-			for key, changes := range m.mergeTracerResult(result) {
-				address := common.HexToAddress(key)
-				err := m.updateAccount(stateTrie, root, address, changes)
-				if err != nil {
-					log.Error("Failed to apply new state", "address", address, "err", err)
-				}
+			acc.Root, err = m.applyStorageChanges(mptStorageTrie, acc.Root, changes)
+			if err != nil {
+				return err
 			}
 		}
 
-		root, err = m.commit(stateTrie, root)
-		if err != nil {
-			return err
-		}
-		if err := m.migratedRef.Update(root, header.Number.Uint64()); err != nil {
+		if err := mptStateTrie.UpdateAccount(addr, acc); err != nil {
 			return err
 		}
 	}
-
 	return nil
 }
 
-func (m *StateMigrator) mergeTracerResult(res prestateTracerResult) map[string]prestateAccount {
-	ret := make(map[string]prestateAccount)
-
-	for addr, pre := range res.Pre {
-		post, modified := res.Post[addr]
-		if modified {
-			acc := prestateAccount{
-				Balance: pre.Balance,
-				Nonce:   pre.Nonce,
-				Code:    post.Code,
-				Storage: make(prestateStorage),
-			}
-			if post.Nonce != nil {
-				acc.Nonce = post.Nonce
+func (m *StateMigrator) applyStorageChanges(mptStorageTrie *trie.StateTrie, storageRoot common.Hash, changes map[common.Hash][]byte) (common.Hash, error) {
+	for hk, val := range changes {
+		preimage, err := m.readZkPreimage(hk)
+		if err != nil {
+			return common.Hash{}, err
+		}
+		slotKey := common.BytesToHash(preimage).Bytes()
+		if (common.BytesToHash(val) == common.Hash{}) {
+			if err := mptStorageTrie.DeleteStorage(common.Address{}, slotKey); err != nil {
+				return common.Hash{}, err
 			}
-			if len(post.Balance) != 0 {
-				acc.Balance = post.Balance
+		} else {
+			trimmed := common.TrimLeftZeroes(common.BytesToHash(val).Bytes())
+			if err := mptStorageTrie.UpdateStorage(common.Address{}, slotKey, trimmed); err != nil {
+				return common.Hash{}, err
 			}
+		}
+	}
 
-			if len(pre.Storage) > 0 || len(post.Storage) > 0 {
-				maps.Copy(acc.Storage, pre.Storage)
-				maps.Copy(acc.Storage, post.Storage)
-				for slot := range acc.Storage {
-					if _, ok := post.Storage[slot]; !ok {
-						acc.Storage[slot] = ""
-					}
-				}
-			}
+	return m.commit(mptStorageTrie, storageRoot)
+}
+
+func (m *StateMigrator) applyNewStateTransition(headNumber uint64) error {
+	start := m.migratedRef.BlockNumber() + 1
+	prevRoot := m.migratedRef.Root()
+	for i := start; i <= headNumber; i++ {
+		log.Info("Apply new state to MPT", "block", i, "prevRoot", prevRoot.TerminalString())
 
-			ret[addr] = acc
+		mptStateTrie, err := trie.NewStateTrie(trie.StateTrieID(prevRoot), m.mptdb)
+		if err != nil {
+			return err
+		}
+
+		destructChanges, accountChanges, storageChanges, err := core.ReadStateChanges(m.db, i)
+		if err != nil {
+			return err
+		}
+		if err := m.applyDestructChanges(mptStateTrie, prevRoot, destructChanges); err != nil {
+			return err
+		}
+		if err := m.applyAccountChanges(mptStateTrie, prevRoot, accountChanges, storageChanges); err != nil {
```

Comment:
I have one more question, what if an account is updated and then deleted in the same block? I think this order will revive the deleted account with previously updated account version, but is it just ok since we don't support SELFDESTRUCT and actual deletion of account is not occurred in hashdb scheme?

Reply:
Good point. The data is created when stateDb.Commit() is processed, so an account in destructChanges (except for accounts marked for deletion, which only affect storages) is not included in accountChanges. Therefore, it’s safe.